src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -1326,6 +1635,285 @@ static void Main()
+        public void DeconstructUsingBaseDeconstructMethod()
+        {
+            string source = @"
+class Base
+{
+    public void Deconstruct(out int a, out int b) { a = 1; b = 2; }
+}
+class C : Base
+{
+    static void Main()
+    {
+        int x, y;
+        (x, y) = new C();
+
+        System.Console.WriteLine(x + "" "" + y);
+    }
+
+    public void Deconstruct(out int c) { c = 42; }
+}
+";
+
+            var comp = CompileAndVerify(source, expectedOutput: "1 2", additionalRefs: new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.Regular.WithTuplesFeature().WithRefsFeature());
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void AssignUsingAmbiguousDeconstruction()
+        {
+            string source = @"
+class Base
+{
+    public void Deconstruct(out int a, out int b) { a = 1; b = 2; }
+    public void Deconstruct(out long a, out long b) { a = 1; b = 2; }
+}
+class C : Base
+{
+    static void Main()
+    {
+        int x, y;
+        (x, y) = new C();
+
+        System.Console.WriteLine(x + "" "" + y);
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib(source, parseOptions: TestOptions.Regular.WithTuplesFeature());
+            comp.VerifyDiagnostics(
+                // (12,18): error CS0121: The call is ambiguous between the following methods or properties: 'Base.Deconstruct(out int, out int)' and 'Base.Deconstruct(out long, out long)'
+                //         (x, y) = new C();
+                Diagnostic(ErrorCode.ERR_AmbigCall, "new C()").WithArguments("Base.Deconstruct(out int, out int)", "Base.Deconstruct(out long, out long)").WithLocation(12, 18)
+                );
+        }
+
+        [Fact]
+        public void DeconstructUsingExtensionMethod()
+        {
+            string source = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        int x;
+        string y, z;
+        (x, (y, z)) = Tuple.Create(1, Tuple.Create(""hello"", ""world""));
+
+        System.Console.WriteLine(x + "" "" + y + "" "" + z);
+    }
+}
+";
+
+            var comp = CompileAndVerify(source, expectedOutput: "1 hello world", additionalRefs: new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.Regular.WithTuplesFeature().WithRefsFeature());
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void NestedDeconstructUsingExtensionMethod()
+        {
+            string source = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        int x;
+        string y, z;
+        (x, (y, z)) = Tuple.Create(1, Tuple.Create(""hello"", ""world""));
+
+        System.Console.WriteLine(x + "" "" + y + "" "" + z);
+    }
+}
+";
+
+            var comp = CompileAndVerify(source, expectedOutput: "1 hello world", additionalRefs: new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.Regular.WithTuplesFeature().WithRefsFeature());
+            comp.VerifyDiagnostics();
+        }
```

Comment:
It's not clear the `Deconstruct` is an extension method in this case, or that it will remain an extension method if the definition of `Tuple` changes. Consider creating an explicit test.